src/ice/mod.rs

```diff
@@ -302,6 +303,64 @@ mod test {
         assert!(a1.poll_transmit().is_none());
     }
 
+    #[test]
+    pub fn migrates_to_new_candidates_after_invalidation_without_timeout() {
+        let _guard = tracing_subscriber::fmt()
+            .with_env_filter("debug")
+            .with_test_writer()
+            .set_default();
+
+        let mut a1 = TestAgent::new(info_span!("L"));
+        let mut a2 = TestAgent::new(info_span!("R"));
+
+        let c1 = host("1.1.1.1:1000", "udp");
+        a1.add_local_candidate(c1.clone());
+        a2.add_remote_candidate(c1.clone());
+
+        let c2 = host("2.2.2.2:1000", "udp");
+        a1.add_remote_candidate(c2.clone());
+        a2.add_local_candidate(c2.clone());
+
+        a1.set_controlling(true);
+        a2.set_controlling(false);
+
+        loop {
+            if a1.state().is_connected() && a2.state().is_connected() {
+                break;
+            }
+            progress(&mut a1, &mut a2);
+        }
+
+        let a1_time = a1.time;
+        let a2_time = a2.time;
+        let new_sock = sock("8.8.8.8:1000");
+
+        let c3 = Candidate::host(new_sock, Protocol::Udp).unwrap();
+        a1.add_local_candidate(c3.clone());
+        a2.add_remote_candidate(c3);
+
+        a1.agent.invalidate_candidate(&c1);
+        a2.agent.invalidate_candidate(&c1);
+
+        loop {
+            let a1_nominated = a1.has_event(
+                |e| matches!(e, IceAgentEvent::NominatedSend { source, .. } if source == &new_sock),
+            );
+            let a2_nominated = a2.has_event(
+                |e| matches!(e, IceAgentEvent::NominatedSend { destination, .. } if destination == &new_sock)
+            );
+
+            if a1_nominated && a2_nominated {
+                break;
+            }
+
+            progress(&mut a1, &mut a2);
+        }
+
+        assert!(a1.time.duration_since(a1_time) < STUN_TIMEOUT);
+        assert!(a2.time.duration_since(a2_time) < STUN_TIMEOUT);
```

Comment:
Without the consideration of remote-candidates in the `invalidate_candidate` function, this assertion fails.